### PR TITLE
Add basic support for `hotwired/turbo` in the back end

### PR DESCRIPTION
**Pull request by m-vo**
This is the first step to add Turbo to the backend. This does not yet enable frames or uses streams at any place but enables drive with instant click/prefetching by default (see below). 

### Status code
The `BackendController` will now return a status code of 422, if a widget was rendered with an error, so that forms still work. For this to work, widgets now set a `_contao_widget_error` attribute on the request in that case, which is evaluated in the controller.

### Assets
There is now a new meta tag `<meta data-turbo-track="reload" content="…">` which contains a hash of the generated stylesheet and script output in the head. If this hash changes, Turbo will issue a full page reload instead of merging. Basically this disables the feature for some pages for now (e.g. the file manager with the Dropzone script added). We can progressively enhance things from here on in follow-up PRs.

### Prefetch
Turbo now - by default - prefetches links on hover. This makes a very nice experience, but we need to disable it for certain things. That is what we are currently doing:

Disable prefetch for…
* … all URLs that contain a `rt` query parameter, so that things like GET requests to delete resources won't be triggered.  
* … everything related to the Sf toolbar.
* … certain menu items (e.g. logout).

We also disable Turbo completely for the `be_login` and `be_password` forms, because it's a completely different structure there.

### User control
If you want to disable Turbo completely, you can do so by setting `Turbo.session.drive = false` in your JS. If you just want to disable prefetch completely, you can add the `<meta name="turbo-prefetch" content="false">` meta tag. 

There is also a fine-grained per-element control using data attributes. See https://turbo.hotwired.dev/handbook/drive#disabling-turbo-drive-on-specific-links-or-forms or https://turbo.hotwired.dev/handbook/drive#prefetching-links-on-hover.

### CODE REVIEW

File: core-bundle/assets/backend.js [hunk 1]
```diff
@@ -19,3 +21,16 @@ application.load(context.keys().map((key) => {
...
+document.documentElement.addEventListener('turbo:before-prefetch', e => {
+    if ((new URLSearchParams(e.target.href)).has('rt')) {
+        e.preventDefault();
+    }
+});
+
+// Make mootools scripts reinitialize themselves
+document.documentElement.addEventListener('turbo:render', () => {
+    window.fireEvent('domready');
+})
```
Comment (fritzmg) [inline core-bundle/assets/backend.js]:
```suggestion
});
```

File: package.json [hunk 2]
```diff
@@ -1,8 +1,11 @@
 {
+    "dependencies": {
+        "@hotwired/stimulus": "^3.2.2",
+        "@hotwired/turbo": "^8.0.4"
+    },
     "devDependencies": {
         "@babel/core": "^7.23.6",
         "@babel/preset-env": "^7.23.6",
-        "@hotwired/stimulus": "^3.2.2",
         "@hotwired/stimulus-webpack-helpers": "^1.0.1",
```
Comment (ausi) [inline package.json:6]:
```suggestion
        "@hotwired/stimulus": "^3.2.2",
        "@hotwired/stimulus-webpack-helpers": "^1.0.1",
        "@hotwired/turbo": "^8.0.4",
```

File: package.json [hunk 3]
```diff
@@ -1,8 +1,11 @@
 {
+    "dependencies": {
+        "@hotwired/stimulus": "^3.2.2",
+        "@hotwired/turbo": "^8.0.4"
+    },
```
Comment (ausi) [inline package.json]:
```suggestion
```

File: core-bundle/assets/backend.js [hunk 4]
```diff
@@ -19,3 +21,16 @@ application.load(context.keys().map((key) => {
...
+// Cancel all prefetch requests that contain a request token
+document.documentElement.addEventListener('turbo:before-prefetch', e => {
+    if ((new URLSearchParams(e.target.href)).has('rt')) {
+        e.preventDefault();
+    }
+});
+
+// Make mootools scripts reinitialize themselves
+document.documentElement.addEventListener('turbo:render', () => {
+    window.fireEvent('domready');
```
Comment (ausi) [inline core-bundle/assets/backend.js]:
```suggestion
    if (!document.body.mooDomreadyFired) {
        document.body.mooDomreadyFired = true;
        window.fireEvent('domready');
    }
```

**Review by fritzmg — CHANGES_REQUESTED** [review]
There are two issues that I noticed:

* When going from any page to the file manager, there will be a JS error regarding DropZone. This is because the DropZone JavaScript is only added there - which is ignored by Turbo by default (without `data-turbo-track="reload"`).
* For some reason the navigation panes in the main nav are getting toggled randomly sometimes, when navigating with Turbo Drive.

**Comment by fritzmg** [comment]
We also need to change all `form.submit()` calls with `form.requestSubmit()` in order for these form submits to be turböd.

File: core-bundle/assets/backend.js [hunk 5]
```diff
@@ -19,3 +21,19 @@ application.load(context.keys().map((key) => {
         return definitionForModuleAndIdentifier(context(key), `contao--${ identifier }`);
     }
 }).filter((value) => value));
+
+/* Turbo support */
+// Cancel all prefetch requests that contain a request token
+document.documentElement.addEventListener('turbo:before-prefetch', e => {
+    if ((new URLSearchParams(e.target.href)).has('rt')) {
```
Comment (m-vo) [inline core-bundle/assets/backend.js]:
```suggestion
    if ((new URLSearchParams(e.target.href)).has('rt') || e.target.closest('.sf-toolbar') !== null) {
```

**Comment by m-vo** [comment]
> We also need to change all form.submit() calls with form.requestSubmit() in order for these form submits to be turböd.

Can we do this in a second step? A quick search found two occurrences (one in `DC_Table`, one in `core.js`), but if we replace the call we also slightly change the behavior as `form.requestSubmit()` does not submit a form that does not validate.

**Comment by fritzmg** [comment]
That should be fine, yes. The form will then just be submitted without Turbo's functionality (i.e. full regular request).

**Comment by m-vo** [comment]
This should be ready for a first review. I updated the description to reflect the changes.

**Review by fritzmg — COMMENTED** [review]
Looks good 👍 

I noticed one issue though: the _Debug mode_ button in the back end also prefetches. We should probably disable prefetch for that menu item (otherwises the debug mode is enabled before you even actually clicked on the button).

Also I noticed that we still have some URLs with request tokens that likely don't actually need one, e.g. the child view, but I'll create a separate issue/PR for that.

**Comment by m-vo** [comment]
> I noticed one issue though: the Debug mode button in the back end also prefetches. We should probably disable prefetch for that menu item (otherwises the debug mode is enabled before you even actually clicked on the button).

Good catch! Fixed in dee9c3f951401f3b527371b7960d89656f2c95a3

**Review by fritzmg — DISMISSED** [review]

**Comment by leofeyer** [comment]
I found two minor issues:

---

Go to the template editor and edit the source of a custom template. Do not save but click on the back button instead. The following error occurs in the browser console:



---

Edit a regular page in the site structure. Do not save but click on the back button instead. The following error occurs in the browser console:



Edit: The same happens if you try to save the page.

---

Any idea how to fix these?

**Comment by leofeyer** [comment]
And when I try to save the template, the following error occurs:

**Comment by m-vo** [comment]
I have the feeling this happens, because we use `document.write()` to load the `ace` scripts. :thinking:  I'll have to investigate.

Edit: Well, kind of. The issue is, that - this way - the script is still loaded even if the element is not present when going to another page. The simplest fix I can imagine would be a return on null. Let's see if the other issues have the same origin.

**Comment by fritzmg** [comment]
If the fix (regarding the template editor) requires too much effort for now we could add

```php
$GLOBALS['TL_HEAD'][] = '<meta name="turbo-visit-control" content="reload">';
```

here

https://github.com/contao/contao/blob/38c088203098c60ae30b36832e6ec78c8e4ecb17/core-bundle/contao/classes/DataContainer.php#L682-L708

 in the meantime.

**Comment by m-vo** [comment]
I added two early returns in a924ba337fa247291808de68d2084c440166b8ef. This is probably not the nicest, but by far the easiest solution I can think of right now.

IMHO the issue can resolve itself when we stop using global JS objects or dynamic script loading with document write.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**